src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.MethodParametersInfo.cs

```diff
@@ -0,0 +1,433 @@
+
+            // 4. Find the default value parameters. Similarly to 3, they should be replaced by the default value.
+            // Example:
+            // Before:
+            // void Caller(int k)
+            // {
+            //     Callee();
+            // }
+            // void Callee(int i = 1, int j = 2)
+            // {
+            //     DoSomething(i, k);
+            // }
+            // After:
+            // void Caller(int k)
+            // {
+            //     DoSomething(1, 2);
+            // }
+            // void Callee(int i = 1, int j = 2)
+            // {
+            //     DoSomething(i, j);
+            // }
+            var operationsWithDefaultValue = allArgumentOperations
+                .WhereAsArray(argument => argument.ArgumentKind == ArgumentKind.DefaultValue);
+
+            // 5. All the remaining arguments, which might includes method call and a lot of other expressions.
+            // Generate a declaration in the caller.
+            // Example:
+            // Before:
+            // void Caller(bool x)
+            // {
+            //     Callee(Foo(), x ? Foo() : Bar())
+            // }
+            // void Callee(int a, int b)
+            // {
+            //     DoSomething(a, b);
+            // }
+            // After:
+            // void Caller(bool x)
+            // {
+            //     int a = Foo();
+            //     int b = x ? Foo() : Bar();
+            //     DoSomething(a, b);
+            // }
+            // void Callee(int a, int b)
+            // {
+            //     DoSomething(a, b);
+            // }
+            var operationsToGenerateFreshVariablesFor = allArgumentOperations
+                .RemoveRange(operationsWithIdentifierArgument)
+                .RemoveRange(operationsWithVariableDeclarationArgument)
+                .RemoveRange(operationsWithLiteralArgument)
+                .RemoveRange(operationsWithDefaultValue)
+                .WhereAsArray(argument => argument.Children.Single().Syntax is TExpressionSyntax);
+
+            var syntaxGenerator = SyntaxGenerator.GetGenerator(document);
+            // There is a special case that should be treated differently. If the parameter is only read once in the method body.
+            // Then use the argument expression to directly replace it.
+            // void Caller(bool x)
+            // {
+            //     Callee(Foo(), Bar())
+            // }
+            // void Callee(int a, int b)
+            // {
+            //     DoSomething(a, b);
+            // }
+            // After:
+            // void Caller(bool x)
+            // {
+            //     DoSomething(Foo(), Bar());
+            // }
+            // void Callee(int a, int b)
+            // {
+            //     DoSomething(a, b);
+            // }
+            var operationsAreReadOnlyOnce = await GetArgumentsReadOnlyOnceAsync(
```

Comment:
```suggestion
            var operationsReadOnlyOnce = await GetArgumentsReadOnlyOnceAsync(
```

or `var operationsThatAreReadOnlyOnce`

Reply:
do you validate that the order didn't hcnage?

either validate that, or doc that you are explicitly not doing htat, and why we think that's ok.

Reply:
Add comments here